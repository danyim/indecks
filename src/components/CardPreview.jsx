import PropTypes from 'prop-types'
import React from 'react'
import { Motion, spring } from 'react-motion'
import Markdown from './Markdown'
import styles from '../styles/components/CardPreview.styl'

const propTypes = {
  title: PropTypes.string,
  answer: PropTypes.string
}

const defaultProps = {
  title: '',
  answer: ''
}

class CardPreview extends React.Component {
  constructor(props) {
    super(props)

    this.togglePreview.bind(this, this.togglePreview)

    this.state = {
      display: false
    }
  }

  togglePreview() {
    const currState = this.state
    currState.display = !currState.display
    this.setState(currState)
  }

  renderPreview(title, answer) {
    return (
      <Motion style={{ y: spring(this.state.display ? 400 : 0) }}>
        {({ y }) =>
          <div
            className={`${styles['preview-pane']}`}
            style={{
              height: y,
              display: y / 400 > 0.05 ? 'block' : 'none'
            }}
          >
            <Markdown className={`${styles['card-title']}`} text={title} />
            <hr />
            <Markdown className={`${styles['card-answer']}`} text={answer} />
          </div>}
      </Motion>
    )
  }

  render() {
    const { title, answer } = this.props

    return (
      <div className={styles.preview}>
        <button
          type="button"
          className="btn pointer m-t"
          onClick={() => this.togglePreview()}
        >
          Preview [{this.state.display ? '-' : '+'}]
        </button>
        {this.renderPreview(title, answer)}
      </div>
    )
  }
}

CardPreview.propTypes = propTypes
CardPreview.defaultProps = defaultProps

export default CardPreview
