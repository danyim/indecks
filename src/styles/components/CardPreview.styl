@import '../_variables'

.card-title p
  font-size: 2.5rem
  padding: .5rem 0

.card-answer p, li
  font-size: 1.5rem
  line-height: 1.5

.preview-pane
  padding: 1.0rem 2.5rem
  background-color: lighten(lightgrey, 90%)
  border: 1px solid lightgrey
