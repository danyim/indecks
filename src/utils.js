export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

