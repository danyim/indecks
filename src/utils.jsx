import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
