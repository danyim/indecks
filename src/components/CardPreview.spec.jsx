import toJson from 'enzyme-to-json'
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import CardPreview from './CardPreview'

const defaultProps = {
  title: 'Test',
  answer: 'test'
}

describe('CardPreview', () => {
  it('should render self and subcomponents', () => {
    const wrapper = mount(CardPreview, defaultProps)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should display the preview pane when the button is pressed', () => {
    const { wrapper } = setupFull(CardPreview, {
      ...defaultProps
    })

    const button = wrapper.find('button[name="preview-toggle"]')
    const labelTitle = wrapper.find('label[htmlFor="title"]')
    const labelAnswer = wrapper.find('label[htmlFor="answer"]')
    expect(button.exists()).toBe(true)

    button.simulate('click')
    expect(labelTitle.exists()).toBe(true)
    expect(labelAnswer.exists()).toBe(true)

    // This is not a true test because the labels are always visible. The
    // <Motion> component is controlling the height of the container based on
    // state.display
  })
})
