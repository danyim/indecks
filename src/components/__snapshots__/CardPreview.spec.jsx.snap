// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardPreview should render self and subcomponents 1`] = `
<CardPreview
  answer="test"
  title="Test"
>
  <div>
    <button
      className="btn pointer m-t"
      name="preview-toggle"
      onClick={[Function]}
      type="button"
    >
      Preview [
      +
      ]
    </button>
    <Motion
      style={
        Object {
          "y": Object {
            "damping": 26,
            "precision": 0.01,
            "stiffness": 170,
            "val": 0,
          },
        }
      }
    >
      <div
        className="undefined"
        style={
          Object {
            "display": "none",
            "height": 0,
          }
        }
      >
        <label
          className="no-pad"
          htmlFor="title"
        >
          <span>
            Title
          </span>
          <Markdown
            className="undefined"
            handleOnClick={null}
            name="title"
            text="Test"
          >
            <div
              className="undefined"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<p>Test</p>
",
                }
              }
              onClick={null}
            />
          </Markdown>
        </label>
        <hr />
        <label
          className="no-pad"
          htmlFor="answer"
        >
          <span>
            Answer
          </span>
          <Markdown
            className="undefined"
            handleOnClick={null}
            name="answer"
            text="test"
          >
            <div
              className="undefined"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<p>test</p>
",
                }
              }
              onClick={null}
            />
          </Markdown>
        </label>
      </div>
    </Motion>
  </div>
</CardPreview>
`;
